### PR TITLE
Add some globalization recommendations for best practices.

### PR DESCRIPTION
**Pull request by Clockwork-Muse**
## Summary

This issue is the documentation side of attempting to deal with issues submitted to the runtime about globalization data errors.

Most of the original text from the [original runtime PR](https://github.com/dotnet/runtime/pull/98899) was already available in various parts of the existing globalization documentation; so far this PR has only adjusted the best-practices documentation.

Do we still fill we need a small "initial troubleshooting" doc?

cc @jkotas @tarekgh @ericstj 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/extensions/best-practices-for-developing-world-ready-apps.md](https://github.com/dotnet/docs/blob/92013ef24a341896c229715dddca8270d51ad499/docs/core/extensions/best-practices-for-developing-world-ready-apps.md) | [Best practices for developing world-ready applications](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/best-practices-for-developing-world-ready-apps?branch=pr-en-us-40492) |
| [docs/core/extensions/globalization-icu.md](https://github.com/dotnet/docs/blob/92013ef24a341896c229715dddca8270d51ad499/docs/core/extensions/globalization-icu.md) | ["Globalization and ICU"](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/globalization-icu?branch=pr-en-us-40492) |

### CODE REVIEW

**Comment by Clockwork-Muse** [comment]
Status checker is complaining about 2 remaining absolute links, but neither link is part of this repo.  What's the way I should handle it?  Still make it relative, assuming that those links would normally be part of the same root regardless?

**Comment by IEvangelist** [comment]
> Status checker is complaining about 2 remaining absolute links, but neither link is part of this repo. What's the way I should handle it? Still make it relative, assuming that those links would normally be part of the same root regardless?

Anything that has the `learn.microsoft.com` host should be relative link instead.

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 1]
```diff
@@ -42,6 +42,11 @@ This section describes the best practices to follow when developing world-ready
...
 
+1. For any element being used for interchange (eg, a field in a JSON document in an API call) or storage, use the <xref:System.Globalization.CultureInfo?displayProperty=InvariantCulture>; additionally, you should explicitly specify a roundtrip format (such as the [`"O"`, `"o"` date-time format specifier](../../standard/base-types/standard-date-and-time-format-strings.md#the-round-trip-o-o-format-specifier)).  Although the format strings for the invariant culture are stable and unlikely to change, specifying an explicit format string helps to clarify the intent of your code.
+    - For date/time elements, consider the advice and observations in Jon Skeet's post [_Storing UTC is not a silver bullet_](https://codeblog.jonskeet.uk/2019/03/27/storing-utc-is-not-a-silver-bullet/).
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:46]:
```suggestion
    - For date/time elements, consider the advice and observations of Noda Time author Jon Skeet, who shares valuable insights. For more information, see [Jon Skeet: _Storing UTC is not a silver bullet_](https://codeblog.jonskeet.uk/2019/03/27/storing-utc-is-not-a-silver-bullet/).
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 2]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
 
 1. Note that the uiCulture value can be set to the browser accept language.
 
+1. For applications that are distributed, allow zero-downtime updates (eg, Azure container applications), or similar you must plan for situations where there may be multiple instances of the application with different format rules or other culture data, most relevantly timezone rules.
+    - If your application deployment includes a database, remember that the database will have its own globalization rules.  In most cases you should avoid performing any globalization-related functions in the database.
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:98]:
```suggestion
    - If your application deployment includes a database, remember that the database will have its own globalization rules. In most cases you should avoid performing any globalization-related functions in the database.
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 3]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
...
+    - If your application deployment includes a client application or web frontend using client globalization resources, assume that the client resources differ from the resources available to your server.  Consider exclusively performing globalization functions on the client.
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:99]:
```suggestion
    - If your application deployment includes a client application or web frontend using client globalization resources, assume that the client resources differ from the resources available to your server. Consider exclusively performing globalization functions on the client.
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 4]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
...
+
+## Recommendations for robust testing
+
+1. To make dependencies more explicit and testing potentially easier and parallelizable, you _SHOULD CONSIDER_ explicitly passing culture-relevant settings, such as `CultureInfo` parameters, to methods that perform formatting, and `TimeZoneInfo` to methods that work with dates and times.  You **MUST** use <xref:System.TimeProvider> or a similar type when retrieving the time.
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:103]:
```suggestion
1. To make dependencies more explicit and testing potentially easier and parallelizable, you _should consider_ explicitly passing culture-relevant settings, such as `CultureInfo` parameters, to methods that perform formatting, and `TimeZoneInfo` to methods that work with dates and times. You _should also_ use <xref:System.TimeProvider> or a similar type when retrieving the time.
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 5]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
...
+
+1. For most tests, you _SHOULD NOT_ explicitly validate the exact output of a given formatting operation or the exact offset of a timezone.  Formatting and timezone data may change at any time, and may differ between two otherwise identical instances of an operating system (and potentially different processes on the same machine).  Relying on an exact value makes tests brittle.
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:105]:
```suggestion
1. For most tests, you _shouldn't_ explicitly validate the exact output of a given formatting operation or the exact offset of a time zone. Formatting and time zone data may change at any time and may differ between two otherwise identical instances of an operating system (and potentially different processes on the same machine). Relying on an exact value makes tests brittle.
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 6]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
...
+    - Generally, validating that some output was received will be sufficient (eg, non-empty strings when formatting).
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:106]:
```suggestion
    - Generally, validating that some output was received will be sufficient (for example, non-empty strings when formatting).
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 7]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
...
+    - For some data elements and formats, validating that the data parses to the input value may be used instead (roundtripping).  Care needs to be taken for cases where fields are dropped (eg, year for some date-related fields) or the value truncated or rounded (such as for floating-point output).
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:107]:
```suggestion
    - For some data elements and formats, validating that the data parses to the input value may be used instead (roundtripping). Care needs to be taken for cases where fields are dropped (for example, year for some date-related fields) or the value truncated or rounded (such as for floating-point output).
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 8]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
...
+    - If you have explicit requirements to validate all localized format output, you _SHOULD CONSIDER_ creating and using a custom culture during test setup.  For most simple cases this can be done by instantiating a `CultureInfo` object via `new CultureInfo(..)` and setting the `DateTimeFormat` and `NumberFormat` properties, and can be entirely stable.  For more complicated cases, subclassing the type allows overriding additional properties.  There are potential additional benefits to this, such as enabling [pseudolocalization](https://learn.microsoft.com/en-us/globalization/methodology/pseudolocalization) via resource files.
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:108]:
```suggestion
    - If you have explicit requirements to validate all localized format output, you _should consider_ creating and using a custom culture during test setup. For most simple cases this can be done by instantiating a `CultureInfo` object with its constructor `new CultureInfo(..)` and setting the `DateTimeFormat` and `NumberFormat` properties. For more complicated cases, subclassing the type allows overriding additional properties. There are potential additional benefits to this, such as enabling [pseudolocalization](/globalization/methodology/pseudolocalization) with resource files.
```

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 9]
```diff
@@ -89,6 +94,20 @@ This section describes the best practices to follow when developing world-ready
...
+    - If you have explicit requirements to validate the results of all date/time operations, you _SHOULD CONSIDER_ creating and using a custom `TimeZoneInfo` instance during test setup.  There are potential additional benefits to this, such as enabling stable testing of certain edge cases (eg, changes to DST rules).
```
Comment (IEvangelist) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:109]:
```suggestion
    - If you have explicit requirements to validate the results of all date/time operations, you _should consider_ creating and using a custom `TimeZoneInfo` instance during test setup. There are potential additional benefits to this, such as enabling stable testing of certain edge cases (for example, changes to DST rules).
```

**Review by IEvangelist — APPROVED** [review]
These are good additions, thank you! I have some nits and suggestions for improvement, after that I think we can :shipit:.

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 10]
```diff
@@ -42,6 +42,11 @@ This section describes the best practices to follow when developing world-ready
...
+    - For date/time elements, consider the advice and observations in Jon Skeet's post [_Storing UTC is not a silver bullet_](https://codeblog.jonskeet.uk/2019/03/27/storing-utc-is-not-a-silver-bullet/).
+
+1. Globalization data [is not stable](https://learn.microsoft.com/en-us/archive/blogs/shawnste/culture-data-shouldnt-be-considered-stable-except-for-invariant), and you should write your application and its tests with this in mind.  It is updated several times a year through host OS channels on all supported platforms.  This data is not maintained by or distributed with the runtime, nor (outside of legacy Windows applications) directly by the host OS.
```
Comment (jkotas) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:48]:
```suggestion
1. Globalization data [is not stable](https://learn.microsoft.com/en-us/archive/blogs/shawnste/culture-data-shouldnt-be-considered-stable-except-for-invariant), and you should write your application and its tests with this in mind.  It is updated several times a year through host OS channels on all supported platforms.  This data is typically not distributed with the runtime.
```
For example, iOS and wasm shipped the globalization data too, it has changed recently with hybrid globalization. I would not into enumerating all past and present cases where the globalization data ships with the runtime or the application.

**Comment by Clockwork-Muse** [comment]
@jkotas @IEvangelist - 
I have a section that reads:
> #### "The data for a culture isn't what it should be!"
> The ICU libraries use the [Unicode Common Locale Data Repository (CLDR)](https://cldr.unicode.org/) as a datasource.  This data is not maintained by the runtime SDK, nor (outside of legacy Windows applications) by the host OS.  If you believe there is an error in the data, please follow [the CLDR directions on submitting a ticket](https://github.com/unicode-org/cldr/blob/main/docs/requesting_changes.md).

Some of the issues that get reported on the main runtime repo are about this sort of thing.    I feel like I maybe want to point people at the CLDR for this sort of thing, but I'm not quite sure where.  I can add it to the runtime bug template (or a bot could auto-reply), or as a footnote to one of the globalization files here.  Thoughts?

File: docs/core/extensions/best-practices-for-developing-world-ready-apps.md [hunk 11]
```diff
@@ -42,6 +42,11 @@ This section describes the best practices to follow when developing world-ready
...
+    - For date/time elements, consider the advice and observations of Noda Time author Jon Skeet, who shares valuable insights. For more information, see [Jon Skeet: _Storing UTC is not a silver bullet_](https://codeblog.jonskeet.uk/2019/03/27/storing-utc-is-not-a-silver-bullet/).
+
+1. Globalization data [is not stable](/archive/blogs/shawnste/culture-data-shouldnt-be-considered-stable-except-for-invariant), and you should write your application and its tests with this in mind. It's updated several times a year through host OS channels on all supported platforms. This data is typically not distributed with the runtime.
```
Comment (jkotas) [inline docs/core/extensions/best-practices-for-developing-world-ready-apps.md:48]:
```suggestion
1. Globalization data [is not stable](https://devblogs.microsoft.com/i18n/culture-data-shouldnt-be-considered-stable/), and you should write your application and its tests with this in mind. It's updated several times a year through host OS channels on all supported platforms. This data is typically not distributed with the runtime.
```
This is an updated version of the same blog post from the same author.

**Comment by jkotas** [comment]
> Some of the issues that get reported on the main runtime repo are about this sort of thing. I feel like I maybe want to point people at the CLDR for this sort of thing, but I'm not quite sure where. I can add it to the runtime bug template (or a ███ could auto-reply), or as a footnote to one of the globalization files here. Thoughts?

Linking to CLDR as the data source in the docs is good idea. I would avoid duplicating instruction for how to file tickets against CLDR in .NET docs. The instructions for that are not that hard to find in the CLDR page. I think it is a good idea to make people slow down and think a bit more before they file CLDR issue to avoid creating noise in the CLDR issue tracker.

I do not think we need to automate this. It is good enough to have a response like
"This globalization issue is by design. Please check <insert the link to the docs here> for more information." that a human uses to close the issue. (Some of our feature teams have a catalog of standard responses like this that they use to close frequently filled by-design issues.)

File: docs/core/extensions/globalization-icu.md [hunk 12]
```diff
@@ -26,6 +26,9 @@ Before .NET 5, the .NET globalization APIs used different underlying libraries o
 
 Starting with .NET 5, developers have more control over which underlying library is used, enabling applications to avoid differences across platforms.
 
+> [!NOTE]
```
Comment (Clockwork-Muse) [inline docs/core/extensions/globalization-icu.md:29]: @jkotas - Added this note.

**Review by jkotas — APPROVED** [review]

File: docs/core/extensions/globalization-icu.md [hunk 13]
```diff
@@ -26,6 +26,9 @@ Before .NET 5, the .NET globalization APIs used different underlying libraries o
...
+> The culture data that drives the behavior of the ICU library is usually maintained by the [Common Locale Data Repository (CLDR)](https://cldr.unicode.org/), not by the runtime or the host OS.
```
Comment (tarekgh) [inline docs/core/extensions/globalization-icu.md:30]: >not by the runtime or the host OS. [](http://example.com/codeflow?start=159&length=34)

Actually this is not accurate as Windows maintain ICU versions which customized a little to fit Windows requirements.  It is possible can run into some data that is not exactly matching CLDR.

Maybe saying `not by the runtime `  and delete the part `or the host OS.` be reasonable?

**Review by tarekgh — APPROVED** [review]
